### DISCUSSION TITLE
Add hardware IP

### DISCUSSION THREAD

**Discussion by btwbtw01** [Xilinx/finn#385] [Q&A] [answered]
Can I add hardware IP into FINN by using Vivado HLS? My understanding of Vivado HLS is writing C code and synthesis, then combine Vivado PS block to generate bistream. The tutorial seems like using python to generate custom IP, and there is limit for generating custom IP. Am I right? If I generate bistream, how could I deploy on FPGA by PYNQ

**Comment by maltanar** [comment]
Hi @btwbtw01 , this is indeed possible but is considered an advanced use-case and we don't have a tutorial for this just yet. There was a previous Gitter discussion on this [here](https://gitter.im/xilinx-finn/community?at=608fc87131d2306a0e8a7ea1) but I'm copying (with some edits) the key steps here, using the recently-added upsampling layer support as a running example:

* add new HLS class in finn-hlslib for new operator ([example](https://github.com/Xilinx/finn-hlslib/blob/master/upsample.hpp))+ HLS unit test ([example](https://github.com/Xilinx/finn-hlslib/blob/master/tb/upsample_tb.cpp))
* add new HLSCustomOp subclass in FINN compiler matching this HLS function ([example](https://github.com/Xilinx/finn/blob/dev/src/finn/custom_op/fpgadataflow/upsampler.py))
* (only needed for some layers) additional transformations in the FINN compiler for data layout conversion ([example](https://github.com/Xilinx/finn/blob/dev/src/finn/transformation/streamline/absorb.py#L558))
* add new transformation `convert_to_hls` to FINN compiler for translating the relevant ONNX layer into the HLSCustomOp instance ([example](https://github.com/Xilinx/finn/blob/dev/src/finn/transformation/fpgadataflow/convert_to_hls_layers.py#L239)) 
* add unit / integration test into FINN compiler to make sure everything works ([example](https://github.com/Xilinx/finn/blob/dev/tests/fpgadataflow/test_fpgadataflow_upsampler.py#L130))

**Reply by btwbtw01** [reply]
Thanks for your response. The hardware IP is like compression and decompression. I just need to compress the activation to store in DRAM and decompress the activation to run convolution.

**Comment by btwbtw01** [comment]
@maltanar Can I just write HLS functions and wrap it up, so I don't have to add CustomOp subclass in FINN compiler?
All I nned to do is add new HLS class in finn-hlslib for new operator
Can I do this without adding additional transformations?
![image](https://user-images.githubusercontent.com/85877169/151405703-accba68f-911d-4595-97fb-9a40eef408ad.png)

**Reply by heborras** [reply]
Hi @btwbtw01,
if I understand correctly you want to have a layer be a convolutional layer. In FINN this is represented by two distinct HLS functions. One for the im2col operation and one for the weight multiplication. So if you want to run the decompression and compression before and after these two functions, then you will need to insert new custom ops there. And the insertion would need to be handled by a transformation.
You could in theory also wrap everything and insert it as such, but that would likely make things more complicated, since one would still need a new custom op along with insertion transformation, while additionally discarding existing functionality for inserting the convolution operation.

So, personally I would recommend to go with creating new ops and transformations for the compression and decompression. The links from Yaman should be very helpful for this: https://github.com/Xilinx/finn/discussions/385#discussioncomment-1459498

**Reply by btwbtw01** [reply]
Hi, @HenniOVP 
Thank you for your answer. Should I need to add custom ops to ONNX graph?
![image](https://user-images.githubusercontent.com/85877169/153351240-d4ce36c5-8028-429b-a5dd-39a9eb9597b2.png)
By the way, I want to ask that what are the difference between step 3 and step 4. Add transformation means that we have to modify the FINN compiler. Am I right?

**Reply by heborras** [reply]
> Hi, @HenniOVP Thank you for your answer. Should I need to add custom ops to ONNX graph? 

Yes, since each HLSCustomOp in FINN is represented by a Python Class and by an ONNX node at the same time you will need to insert them into the graph at some point before synthesis.

> By the way, I want to ask that what are the difference between step 3 and step 4. 

Yes, adding transformations and custom op's means that you will have to make changes to FINN.

> Add transformation means that we have to modify the FINN compiler. Am I right?

Well, the difference between step 3 and 4 is that, step 3 handles general network preparation, which may not be needed in your case and step 4 converts the normal ONNX operations to their HLS counterparts, this is probably the place where you would want to insert your compression/decompression.

**Reply by btwbtw01** [reply]
I have asked the question about stiching IP before https://gitter.im/xilinx-finn/community?at=610529ab7331d202b5dd9d00, it says that we can use stitched IP and add the additional Vivado/Vits HLS IP blocks. Maybe I can try this method and don't have to modify 
![image](https://user-images.githubusercontent.com/85877169/153535550-9ec7cca4-ffcf-4653-a498-293449e25327.png)
I have asked questions about stiched IP and MO before. https://github.com/Xilinx/finn/discussions/516 https://github.com/Xilinx/finn/discussions/520
Sorry for asking a lot of questions, and thank you for responding. I just want to understand the framework of FINN to use in my project.

**Reply by heborras** [reply]
Well, as Yaman said you can certainly add things behind or in front of an IP-Block generated by FINN. However, since FINN doesn't create MO architectures out of the box, the additions will be at the beginning and end of the neural network. And as I understood you want to have the compression / decompression between the layers, so i doubt that this approach will do what you want.

I also just remembered, that you could potentially build an MO architecture yourself by making use of run-time writable weights, which FINN supports. However, this will be a lot of additional work, so I'm not sure if it's a viable option for you.

**Reply by btwbtw01** [reply]
Hi, @HenniOVP 
Are the steps that Yaman mentioned about adding custom IP same as https://github.com/Xilinx/finn/blob/main/notebooks/advanced/2_custom_op.ipynb?

**Reply by heborras** [reply]
Hi,
the steps from Yaman are a bit more extensive, but in general they are the similar to the ones in the notebook.

**Reply by btwbtw01** [reply]
Hi, @HenniOVP 
What is the function of custom op? Is setting the folding factors and specify properties of these nodes? How the example calls these python codes to create custom op? 
![image](https://user-images.githubusercontent.com/85877169/156738461-e4cc01ca-5c91-4f51-b3e5-17175e49e964.png)
I guess I have to add compression nodes between the hls layer after Creating a Dataflow Partition and during Adjusting the Parallelism
![image](https://user-images.githubusercontent.com/85877169/156738490-af5ddc66-00df-4b37-a2f7-2958e13f6b1f.png)

**Reply by btwbtw01** [reply]
According to notebooks/advanced/2_custom_op.ipynb, the most important function to add custom ops is get_nodeattr_types and execute_node. Am I right? In the model transform, we will first convert the graph to HLS, than execute custom ops. In Yaman example, it will execute transformation convert_to_hls first, then execute HLSCustomOp subclass. Is that correct?

**Reply by btwbtw01** [reply]
Hi, @HenniOVP 
May I ask how the FINN compiler generate FIFO between two computing arrays? Can I have two FIFO between two computing arrays like this?
![image](https://user-images.githubusercontent.com/85877169/157879861-f3d9ccf9-b1d1-4936-873d-7af959672c53.png)
Thanks!!

**Reply by JPPalacios** [reply]
Hi @btwbtw01,

Were you able to create custom HLS Ops? If so, could you provide some details on how you went about it? I am looking for something similar but before/after.

Thanks

**Comment by auphelia** [comment, marked as answer]
Adding this link here, since the custom op integration has slightly changed in the latest release: https://github.com/Xilinx/finn/discussions/1026#Refactoring-of-Custom-Operator-Infrastructure

And the additional discussion here: https://github.com/Xilinx/finn/discussions/1160 (thanks @JPPalacios and @iksnagreb)
